### PR TITLE
fix(core): recognize symlinked agent files in loadAgentsFromDirectory

### LINKED ISSUE #20079: ~/.gemini/agents/filename.md is not recognized as an agent if filename.md is a symlink
### What happened?

When ~/.gemini/agents/filename.md is a symlink, it does not get recognized as an agent

### What did you expect to happen?

That you could have a symlink in ~/.gemini/agents/ and it would get recognized as a subagent.

### Client information

<details>
<summary>Client Information</summary>

Run `gemini` to enter the interactive CLI, then run the `/about` command.

```console
> /about
# paste output here
```

</details>

### Login information

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by BryanBradfo**
Node.js `Dirent.isFile()` returns `false` for symbolic links, even when the link target is a regular file. This caused `loadAgentsFromDirectory()` to silently skip any `.md` agent definition that was a symlink in `~/.gemini/agents/` or `.gemini/agents/`.

## Summary

Symlinked `.md` agent files in `~/.gemini/agents/` or `.gemini/agents/` were silently ignored. Added `entry.isSymbolicLink()` to the directory entry filter so symlinks to valid agent files are discovered and loaded.

## Details

The filter in `loadAgentsFromDirectory()` (`agentLoader.ts:501`) only checked `entry.isFile()`, which per Node.js docs evaluates the dirent type — not the resolved target. Symlinks have their own dirent type and fail this check.

Adding `entry.isSymbolicLink()` is sufficient because `fs.readFile()` (used downstream) already follows symlinks transparently. No other code paths needed changes.

## Related Issues

Fixes #20079

## How to Validate

1. `npm run preflight` — all tests pass
2. Create a symlinked agent:
```bash
cat > /tmp/agent.md << 'EOF'                           
---                                                    
name: test-agent
description: Test
---
Hello
EOF
ln -sf /tmp/agent.md ~/.gemini/agents/test-agent.md
```
3. npm run build && npm run start
4. Type @ in the CLI — test-agent should appear in the list

## Pre-Merge Checklist

- [] Updated relevant documentation and README (if needed)
- [x] Added/updated tests (if needed)
- [ ] Noted breaking changes (if any)
- [ ] Validated on required platforms/methods:
  - [ ] MacOS
    - [ ] npm run
    - [ ] npx
    - [ ] Docker
    - [ ] Podman
    - [ ] Seatbelt
  - [ ] Windows
    - [ ] npm run
    - [ ] npx
    - [ ] Docker
  - [x] Linux
    - [x] npm run
    - [ ] npx
    - [ ] Docker

### CODE REVIEW

File: packages/core/src/agents/agentLoader.ts [hunk 1]
```diff
@@ -498,7 +498,7 @@ export async function loadAgentsFromDirectory(
 
   const files = dirEntries.filter(
     (entry) =>
-      entry.isFile() &&
+      (entry.isFile() || entry.isSymbolicLink()) &&
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts:514]: Thanks for catching this security issue! I have updated the logic to address the Path Traversal vulnerability.
The code now uses fs.realpath() to resolve the symlink target, and I added isSubpath() validation to ensure the resolved path stays strictly within the intended agents directory. I also added two new test cases to properly catch out-of-bounds and dangling symlinks as AgentLoadErrors.

File: packages/core/src/agents/agentLoader.test.ts [hunk 2]
```diff
@@ -354,6 +354,28 @@ Hidden`,
...
+        path.join(tempDir, 'original.md'),
+        path.join(tempDir, 'symlinked.md'),
+      );
+
+      const result = await loadAgentsFromDirectory(tempDir);
+      expect(result.errors).toHaveLength(0);
+      expect(result.agents).toHaveLength(2);
+      const names = result.agents.map((a) => a.name);
+      expect(names).toContain('original-agent');
+    });
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.test.ts]: Thank you! I've also added the additional tests for the security edge cases

File: packages/core/src/agents/agentLoader.test.ts [hunk 3]
```diff
@@ -354,6 +354,66 @@ Hidden`,
...
+      await fs.symlink(
+        path.join(tempDir, 'original.md'),
+        path.join(tempDir, 'symlinked.md'),
+      );
+
+      const result = await loadAgentsFromDirectory(tempDir);
+      expect(result.errors).toHaveLength(0);
+      expect(result.agents).toHaveLength(2);
+      const names = result.agents.map((a) => a.name);
+      expect(names).toContain('original-agent');
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.test.ts]: Implemented deduplication via a seenPaths Set that tracks resolved paths from fs.realpath. The test now expects toHaveLength(1) and confirms only one agent is loaded when both a file and its symlink exist in the directory.

File: packages/core/src/agents/agentLoader.ts [hunk 4]
```diff
@@ -498,15 +499,26 @@ export async function loadAgentsFromDirectory(
...
       !entry.name.startsWith('_') &&
       entry.name.endsWith('.md'),
   );
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts:517]: I removed the redundant entry.isFile() && and added deduplication using a seenPaths Set keyed on resolved paths from fs.realpath. Each unique agent file is now loaded only once, even when accessed via multiple symlinks. Broken symlinks and symlinks pointing outside the agents directory are also handled with appropriate error reporting.

File: packages/core/src/agents/agentLoader.ts [hunk 5]
```diff
@@ -512,14 +511,47 @@ export async function loadAgentsFromDirectory(
...
     try {
+      let realPath: string;
+      try {
+        realPath = await fs.realpath(filePath);
+      } catch (e) {
+        if (entry.isSymbolicLink()) {
+          result.errors.push(
+            new AgentLoadError(filePath, 'Symlink target does not exist'),
+          );
+          continue;
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts:535]: the startsWith check is now replaced with path.relative containment, exactly as suggested. Also added a dedicated test for the prefix-bypass scenario (e.g., agentsDir + '-evil').

File: packages/core/src/agents/agentLoader.ts [hunk 6]
```diff
@@ -512,14 +511,47 @@ export async function loadAgentsFromDirectory(
...
-      entry.isFile() &&
+      (entry.isFile() || entry.isSymbolicLink()) &&
       !entry.name.startsWith('_') &&
       entry.name.endsWith('.md'),
   );
 
+  const resolvedDir = path.resolve(dir);
+  const seenPaths = new Set<string>();
+
   for (const entry of files) {
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts:522]: The fs.realpath error handling ensures broken symlinks produce a clear AgentLoadError rather than crashing the loader, so users get actionable diagnostics.

File: packages/core/src/agents/agentLoader.ts [hunk 7]
```diff
@@ -512,14 +511,47 @@ export async function loadAgentsFromDirectory(
...
+          result.errors.push(
+            new AgentLoadError(filePath, 'Symlink target does not exist'),
+          );
+          continue;
+        }
+        throw e;
+      }
+
+      // Security: reject symlinks pointing outside the agents directory
+      if (entry.isSymbolicLink() && !realPath.startsWith(resolvedDir)) {
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts]: Thanks! The seenPaths dedup ensures that if multiple symlinks resolve to the same file, the agent is only loaded once, avoiding duplicate registrations and wasted parsing.

File: packages/core/src/agents/agentLoader.ts [hunk 8]
```diff
@@ -512,14 +511,51 @@ export async function loadAgentsFromDirectory(
...
+      try {
+        realPath = await fs.realpath(filePath);
+      } catch (e) {
+        if (entry.isSymbolicLink()) {
+          result.errors.push(
+            new AgentLoadError(filePath, 'Symlink target does not exist'),
+          );
+          continue;
+        }
+        throw e;
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply packages/core/src/agents/agentLoader.ts:537]: Thanks for the suggestion! Now distinguishing ENOENT (broken symlink) from other failures like ELOOP or EACCES, with isNodeError imported from ../utils/errors.js

**Comment by abhipatel12** [comment]
Hey thanks for the PR. This issue is marked for maintainers only, so going to close out the PR.
